src/task.rs

```diff
@@ -0,0 +1,421 @@
+//! loop, or generally becomes CPU bound for a sustained period of time, other
+//! tasks in the system are not starved of resources, and system latency does
+//! not fall off a cliff.
+//!
+//! Tasks are **isolated** from each other and share no memory. They communicate
+//! by **message passing** over a readable and writable streams. This means that
+//! if one task's local state is corrupted, it doesn't immediately infect all
+//! other tasks' states that are communicating with it. The corrupted task can
+//! be killed by its supervisor and then respawned with the last known good
+//! state or with a clean slate.
+//!
+//! Tasks are **lightweight**. In order to facilitate let-it-fail error handling
+//! coupled with supervision hierarchies, idle tasks have little memory
+//! overhead, and near no time overhead.
+
+use super::{Error, ErrorKind, Result, StarlingHandle, StarlingMessage};
+use futures::{self, future, Async, Future, Sink};
+use futures::sync::oneshot;
+use futures_cpupool::CpuFuture;
+use futures_mpsc::{self, Receiver, Sender};
+use js;
+use js::jsapi;
+use js::rust::Runtime as JsRuntime;
+use std::fmt;
+use std::os;
+// use std::panic;
+use std::path;
+use std::ptr;
+use std::thread;
+use tokio_core::reactor::Core as TokioCore;
+
+/// A unique identifier for some `Task`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct TaskId(thread::ThreadId);
+
+impl From<TaskId> for thread::ThreadId {
+    fn from(task_id: TaskId) -> Self {
+        task_id.0
+    }
+}
+
+/// A `Task` is a JavaScript execution thread.
+///
+/// A `Task` is not `Send` nor `Sync`; it must be communicated with via its
+/// `TaskHandle` which is both `Send` and `Sync`.
+///
+/// See the module level documentation for details.
+pub(crate) struct Task {
+    handle: TaskHandle,
+    receiver: Receiver<TaskMessage>,
+    global: js::heap::Heap<*mut jsapi::JSObject>,
+    runtime: JsRuntime,
+    starling: StarlingHandle,
+    js_file: path::PathBuf,
+    parent: Option<TaskHandle>,
+    state: State,
+}
+
+/// ### Constructors
+///
+/// There are two kinds of tasks: the main task and child tasks. There is only
+/// one main task, and there may be many child tasks. Every child task has a
+/// parent whose lifetime strictly encapsulates the child's lifetime. The main
+/// task's lifetime encapsulates the whole system's lifetime.
+///
+/// There are two different constructors for the two different kinds of tasks.
+impl Task {
+    /// Spawn the main task.
+    ///
+    /// The lifetime of the system is tied to this task. If it exits
+    /// successfully, the process will exit 0. If it fails, then the process
+    /// will exit non-zero.
+    pub fn spawn_main(
+        starling: StarlingHandle,
+        js_file: path::PathBuf,
+    ) -> Result<(TaskHandle, thread::JoinHandle<()>)> {
+        let (send_handle, recv_handle) = oneshot::channel();
+        let starling2 = starling.clone();
```

Comment:
Any reason this cannot be inlined in the `create_main` call directly? (Mainly asking because I dislike the name.)

Reply:
I was trying to separate the concepts of spawning a thread and creating a task, since we intend to have them be different things down the line.

Reply:
I don't fully understand why that needs the `starling2` local, but given that this is very much a nit, I also don't feel I have to :)

Reply:
Ah, I didn't realize you were specifically referring to `starling2`. This is because the first `starling` gets moved into the thread, and `StarlingHandle` is not `Copy` so we have to explicitly clone it into a new variable.

Reply:
Yeah, all I meant was to not have this local and do `starling.clone()` inline in the function call. Clearly this isn't in any way important, though.